pageserver/src/tenant/storage_layer/image_layer.rs

```diff
@@ -623,6 +623,9 @@ impl ImageLayerWriterInner {
                 lsn: self.lsn,
             },
         );
+        if final_path.exists() {
+            warn!("Overwrite existed image layer {}", final_path.display());
+        }
```

Comment:
I think we should prevent this case earlier. How do we end up creating a duplicate image layer?

Reply:
This warning was added just to verify that no redundant image layers are created.
It is the problem originally detected by you: generated "wanted" image layer doesn't allow immediately delete old layers because it is newer than PITR boundar and some once gain will be included in wanted list.
I have added check for presence of such layer and you asked to add test for this scenario.
This warning is part of test test: before this check was added such warnings are actually reported by test_gc_feedback.py

Reply:
Oh I see. So we don't currently generate such duplicate layers, but if we did, you'd now get this warning.

Let's make it a hard error. I'm worried that the rest of the code might not handle that case well, so it would be better to fail early.

It would be nice to detect if the file exists when we do the 'rename', instead of checking for it beforehand. In case there's a race condition where two tasks try to create the same image layer at the same time. We only have a single task that creates image layers at the moment, but still. There's no way to perform a rename that would not overwrite an existing file, though. There is `renameat2()` but it's Linux-specific and there's no std rust wrapper for it.

Perhaps we should add a check in the caller, before calling `ImageLayerWriter::finish`, to check that there's no duplicate layer in the layer map already? Checking for a pre-existing file on local filesystem doesn't catch the case that there's a duplicate file in S3. And same for creating a  new delta layer.

Reply:
First of all presence of duplicated image layers seems to be not so fatal.
At least our examples normally work when there are such layers.
We now accept overlapping layers. The only requirement is that intersected ranges are identical.
This rule is satisfied for duplicated image layers. May be there will be some problems with S3 download - it is what I not checked.

But I agree with you that it should not happen and there are no any valid reasons which can cause generation of duplicated layers. Detecting  duplicates at layer map level will require some changes in Bojan's LayerMap. I wonder if we should prohibit just image layer duplicates and delta layer dulicates as well?

Reply:
> Detecting duplicates at layer map level will require some changes in Bojan's LayerMap.

Hmm, yes, need to add a new function that checks if a layer with given LSN and key range exists. That's not difficult, is it? There's already a 'replace' function that has to do the same work.

> I wonder if we should prohibit just image layer duplicates and delta layer dulicates as well?

Yeah, we should prohibit both.

Reply:
Done